proto/substrait/algebra.proto

```diff
@@ -340,6 +340,22 @@ message ExchangeRel {
   }
 }
 
+// The expand relation generates multiple rows per input row (denoted by
+// groupings). No ordering is applied to the results.
+message ExpandRel {
+  RelCommon common = 1;
+  Rel input = 2;
+
+  // A list of expression grouping that the aggregation measured should be calculated for.
+  repeated Grouping groupings = 3;
+
+  message Grouping {
```

Comment:
A few thoughts:
* I didn't see whether there was an answer to my question on whether arbitrary expressions are allowed inside Acero and Spark. Are they?
* I don't think grouping is a good name for what we're doing. We are not grouping anything.
* I don't understand by the comments whether this operator maintains order or not. The comment about order is vague to me.
* I think the pseudo-code put forth about how we calculate output type and output fields is problematic and suggests that we have the wrong representation. It also works opposite of the way aggregate works which makes it less desireable.
* It isn't clear what the conclusion is wrt emit of existing columns. 

As I think more about this I think the grouping pattern just seems wrong given the requirements around same data types, etc. I almost think that the array of values should be transposed. For example:

```
// Duplicates records, possibly switching output expressions between each duplicate.
// Default output is all of the fields declared + one field that int64 type that returns the duplicate_id which is a zero-index ordinal of which duplicate of the original record this corresponds to
message ExpandRel {
  RelCommon common = 1;
  Rel input = 2;
  // tells how many times output records there are for each input record.
  int duplicate_count = 3; // Not strictly necessary
  repeated ExpandField fields = 4;

  message ExapandField {
    Type type = 1;
    oneof field_type {
      // Field that switches output based on which duplicate_id we're outputting
      SwitchingField switching_field = 2;

      // Field that outputs the same value no matter which duplicate id we're on.
      Expression consistent_field = 3;
    } 
  }

  message SwitchingField {
    // array that contains the expression to output per expansion_id
    // must have the same number of expressions as each input record
    // all expressions must be of the same type
    repeated Expression duplicates = 1;
  }

}


```

I think this does a much clearer job of expressing what we're supporting. We're duplicating records. In some fields, each duplicate gets a different expression. In some cases, all duplicates get the same record.

Thoughts?

Reply:
> I didn't see whether there was an answer to my question on whether arbitrary expressions are allowed inside Acero and Spark. Are they?

For Acero, the answer is yes.  I don't know for Spark.  Given all the use cases I am aware of we could probably restrict this to "field refs or literals" but I don't know if that would be useful or not.

> I don't think grouping is a good name for what we're doing. We are not grouping anything.

If it helps, in Acero we use "templates".

> I don't understand by the comments whether this operator maintains order or not. The comment about order is vague to me.

There are a few different styles you can use here.  E.g. given ROW1, ROW2, ROW3 and two duplicates you could have `ROW1D1 ROW1D2 ROW2D1 ROW2D2 ROW3D1 ROW3D2` or you could have `ROW1D1 ROW2D1 ROW3D1 ROW1D2 ROW2D2 ROW3D2`.  Dplyr, for example, in pivot_longer, allows you to specify one or the other.  Acero however, does not actually support either of those for performance reasons (we implement this as a zero-copy operation).

So I interpreted the comment about order as saying that we are not guaranteeing any specific ordering of rows.

> As I think more about this I think the grouping pattern just seems wrong given the requirements around same data types, etc. I almost think that the array of values should be transposed

Your proposal (with `ExpandField`) looks good to me.  I can see how it will simplify the calculation of the output type.  I'd rather not include `duplicate_count` since it should be redundant (each `SwitchingField` should have the same number of `duplicates`).

Reply:
To clarify, I'd rather not include `duplicate_count` but I'd be fine with leaving it in.

Reply:
> Arbitrary expressions allowed in acero

Ok, I figured as much which lead me to propose this with expressions as opposed to subset of items.

>  I'd rather not include duplicate_count 

Yeah, I actually commented it and then uncommmented it. I'm generally also inclined to exclude. If you have duplicate data then you have to specify the behavior when it is inconsistent :)

@baibaichen and @JkSelf , what are your thoughts on my inverted representation?

Reply:
@jacques-n 
I don't understand how `ExpandField` works, it looks complicated to me. I am not clear about your concern, why we need duplcate_id? Back to `Expand`'s semantic：**Apply a number of projects to every input row,  hence we will get mulitple ouput rows for an input row**.  

> I don't think grouping is a good name for what we're doing. We are not grouping anything.

Yes, I prefer to `projections`,  I think grouping us confusing us. 

For other comments:

> I didn't see whether there was an answer to my question on whether arbitrary expressions are allowed inside Acero and Spark. Are they?

Yes, arbitrary expressions.


> I don't understand by the comments whether this operator maintains order or not. The comment about order is vague to me.

Spark uses expression id to refer input's(child plan）Field, while Calcite uses index. So When substrait plan is convertted back to Spark plan,  The field order is very important.

For `ProjectRel`, we **assume** output order is same as its project expression, which is right for Spark `Project`. But, I don't know whether this assumption is always ture for `Expand` or not.

> I think the pseudo-code put forth about how we calculate output type and output fields is problematic and suggests that we have the wrong representation. It also works opposite of the way aggregate works which makes it less desireable.

I think it's OK for now. It's better to add `output`, but we can add it latter once assumption is broken.

> It isn't clear what the conclusion is wrt emit of existing columns.

I don't understand.

Reply:
Here the `ExpandField ` maybe not suite for spark. Because we only get the projection list in [ExpandExec ](https://github.com/apache/spark/blob/cb7d08280623794b238297d2f3de8abdc8b72bdb/sql/core/src/main/scala/org/apache/spark/sql/execution/ExpandExec.scala#L37)and it is hard to distinguish the consistent_field and switching_field.

Reply:
> Here the `ExpandField ` maybe not suite for spark. Because we only get the projection list in [ExpandExec ](https://github.com/apache/spark/blob/cb7d08280623794b238297d2f3de8abdc8b72bdb/sql/core/src/main/scala/org/apache/spark/sql/execution/ExpandExec.scala#L37)and it is hard to distinguish the consistent_field and switching_field.

I think you're saying that it is a two dimensional array. In those cases, you'd create all items switching fields. I don't see a problem.

> I don't understand how ExpandField works, it looks complicated to me

What is unclear or complex? 

I believe this is actually a simplification that more accurately represents the semantics that you desire. In expand: you're producing N output rows for each input row. All output rows have M columns. For each column in M, you do one of two things when outputting a column/field: 
1. duplicate the value same expression N times (consistent_field)
2. use a different value for each of the N rows (switching_field)
 
ExpandField clearly distinguishes between these two behaviors.

This representation also makes a clear hierarchical relationship such that it is clear that all items within a list of switching field must have the same data type.

Reply:
>I think you're saying that it is a two dimensional array. In those cases, you'd create all items as switching fields. I don't see a problem with mapping to/from this representation. Additionally, if you received something that had consistent fields and wanted to convert to Spark representation, it would be a simple matter of duplication the expression for the consistent field across all expansion items.

Yes,  as I talked to @JkSelf offline. we will create all items as switching fields. If you don't think it's a problem, then do it this way.

Reply:
> > I think you're saying that it is a two dimensional array. In those cases, you'd create all items as switching fields. I don't see a problem with mapping to/from this representation. Additionally, if you received something that had consistent fields and wanted to convert to Spark representation, it would be a simple matter of duplication the expression for the consistent field across all expansion items.
> 
> Yes, as I talked to @JkSelf offline. we will create all items as switching fields. If you don't think it's a problem, then do it this way.
@jacques-n Updated the ExpandRel. Do you have any further suggestion? Thanks.